### ISSUE TITLE
[Mandrel 22.3.x] Datasonnet integration tests fail in native mode

### ISSUE THREAD

**Issue by jamesnetherton** [apache/camel-quarkus#4284]
```
Caused by: java.lang.IllegalArgumentException: Class sjsonnet.Expr$Member$Field[] is instantiated reflectively but was never registered.Register the class by adding "unsafeAllocated" for the class in reflect-config.json.
        at org.graalvm.nativeimage.builder/com.oracle.svm.core.graal.snippets.SubstrateAllocationSnippets.arrayHubErrorStub(SubstrateAllocationSnippets.java:345)
        at scala.reflect.ClassTag$GenericClassTag.newArray(ClassTag.scala:149)
        at scala.collection.mutable.ArrayBuilder$ofRef.mkArray(ArrayBuilder.scala:109)
        at scala.collection.mutable.ArrayBuilder$ofRef.resize(ArrayBuilder.scala:114)
        at scala.collection.mutable.ArrayBuilder.ensureSize(ArrayBuilder.scala:38)
        at scala.collection.mutable.ArrayBuilder$ofRef.addOne(ArrayBuilder.scala:119)
        at scala.collection.mutable.ArrayBuilder$ofRef.addOne(ArrayBuilder.scala:103)
        at scala.collection.ArrayOps$WithFilter.map(ArrayOps.scala:90)
        at sjsonnet.Materializer$.toExpr(Materializer.scala:99)
        at com.datasonnet.Mapper.transform(Mapper.scala:245)
        at org.apache.camel.language.datasonnet.DatasonnetExpression.doEvaluate(DatasonnetExpression.java:151)
        at org.apache.camel.language.datasonnet.DatasonnetExpression.evaluate(DatasonnetExpression.java:90)
        ... 46 more
```

**Comment by jamesnetherton** [comment]
Seems this is specific to Mandrel 22.3.x. If I build & run with GraalVM 22.3.0, then the test passes.

**Comment by ppalaga** [comment]
I see this happening also with Quarkus 2.14.2 and Mandrel 22.3.0 in https://github.com/apache/camel-quarkus/actions/runs/3555491649/jobs/5972546804#step:8:8093 Let me try to remove Quarkus upgrade from https://github.com/apache/camel-quarkus/pull/4224

**Comment by ppalaga** [comment]
@galderz @zakkak is the different behavior between GraalVM and Mandrel expected?

I tried to figure out how to implement the hint `adding "unsafeAllocated" for the class in reflect-config.json` given in the error message.

There seem to exist no Quarkus way to register a class for reflection with `unsafeAllocated=true` or I am overseeing something?

Adding 

```
[
    {
        "name": "sjsonnet.Expr$Member$Field[]",
        "unsafeAllocated": true
    },
    {
        "name": "sjsonnet.Expr[]",
        "unsafeAllocated": true
    }
]
```

to extensions/jsonata/runtime/src/main/resources/META-INF/native-image/org.apache.camel.quarkus/camel-quarkus-jsonata/reflect-config.json helps though. I have sent that in https://github.com/apache/camel-quarkus/pull/4343 but I consider it only to be a workaround.

Is this a bug in Mandrel or we need to add unsafeAllocated to ReflectiveClassBuildItem?

**Comment by zakkak** [comment]
> @galderz @zakkak is the different behavior between GraalVM and Mandrel expected?

No, that's interesting!
 
> I tried to figure out how to implement the hint `adding "unsafeAllocated" for the class in reflect-config.json` given in the error message.
> 
> There seem to exist no Quarkus way to register a class for reflection with `unsafeAllocated=true` or I am overseeing something?

AFAIK indeed there is no such option at the moment. 

> Is this a bug in Mandrel or we need to add unsafeAllocated to ReflectiveClassBuildItem?

We need to understand why this is different in Mandrel when compared to GraalVM. Regarding adding `unsafeAllocated` to `ReflectiveClassBuildItem` my guess is that we probably need it, but let's see why this only appears in Mandrel first :)

**Comment by ppalaga** [comment]
> The first thing I notice examining this is that the error is triggered by some scala code. And when it comes to scala it looks like GraalVM CE is using `com.oracle.svm.polyglot.scala.ScalaFeature` which is not available in Mandrel. This could explain why it works with GraalVM and not with Mandrel.
>
> Now regarding how to approach it. Mandrel doesn't support any language other than Java so a fix for this is probably not going to land on Mandrel.

Here the use case is an application written in Java using a library written in Scala. I wonder whether this use case should be supported by Mandrel. WDYT @maxandersen ?

> On the other hand, extending Quarkus to support `unsafeAllocated=true` seems reasonable to me.

Thanks, I have filed https://github.com/quarkusio/quarkus/issues/29922

**Comment by jamesnetherton** [comment]
This was fixed properly back in 3.0.0-M2.
